crates/ruff/src/rules/ruff/rules/mutable_class_default.rs

```diff
@@ -47,42 +51,44 @@ impl Violation for MutableClassDefault {
 
 /// RUF012
 pub(crate) fn mutable_class_default(checker: &mut Checker, class_def: &ast::StmtClassDef) {
-    for statement in &class_def.body {
-        match statement {
-            Stmt::AnnAssign(ast::StmtAnnAssign {
-                annotation,
-                value: Some(value),
-                ..
-            }) => {
-                if is_mutable_expr(value, checker.semantic())
-                    && !is_class_var_annotation(annotation, checker.semantic())
-                    && !is_final_annotation(annotation, checker.semantic())
-                    && !is_immutable_annotation(annotation, checker.semantic())
-                    && !is_dataclass(class_def, checker.semantic())
-                {
-                    // Avoid Pydantic models, which end up copying defaults on instance creation.
-                    if is_pydantic_model(class_def, checker.semantic()) {
-                        return;
-                    }
+    if class_def.body.iter().any(Stmt::is_ann_assign_stmt) {
+        for statement in &class_def.body {
```

Comment:
This changes the runtime of this rule to always iterate over the body twice. I don't know if this is a big issue, considering that classes only have a few top-level statements (it skips over all function bodies). 

However, I don't have a good idea that is guaranteed to be faster:

* Collecting the diagnostics and ignoring them if no assignment statement was seen is probably slower it requires allocating the diagnostics `Vec` and the messages in `DiagnosticKind`. 
* Filtering out the `AnnAssignStmt` and `AnnAssign` statements and storing them in a `Vec`. Running `is_any` and the check only over that sublist. Again, the allocation may be as expensive as iterating over all statements.